Export to Typst/Fletcher diagram

tjbcg opened the pull request:
I've implemented initial support to export to Typst diagrams. Incompatibilities are reported through the incompatibilities metadata object (that is: shortened arrows and k-cells for k > 1)

This still needs some UI adjustments to show the incompatibilities and instructions just like with latex.

I marked this as Draft because I feel like this feature cannot be fully used as-is since I haven't implemented in-browser rendering of the formulas. Looking into how to do that with `typst.ts`.

Feel free to leave comments and to try this out. Diagrams should be readily useable with
```typ
#import "@preview/fletcher:0.5.5" as fletcher: diagram, node, edge
```
at a document's start.

For those wanting to try it, I've got a version hosted [here](https://quiver.theophile.me/)

tjbcg commented:
For a bit of fun here's an example in which I crammed almost every variation of options (except those that are incompatible with Fletcher)

![image](https://github.com/user-attachments/assets/94ab26b1-ff38-42d0-88b6-1d336f62a34c)
![image](https://github.com/user-attachments/assets/dd68eb3b-ab7b-4416-afca-0bb87295a43a)

tjbcg commented:
@varkor Couple questions if you have time for them.

At this point I've implemented Typst rendering, but the (AFAIK only) library is very bare bones so I've had to work around some minor issues. Notably, concurrent renders seem to break the library but I've implemented a queue for the rendering, which is quick enough to work as a permanent solution. Onto real issues:

- Typst.ts loads a 20MB binary in memory (it's the wasm-compiled typst compiler). Would you rather have everything load from the project as is currently the case with KaTeX, at the expense of having a 20M binary added to the directory, or would you rather load the library & binary from a CDN ? Obviously there's no doubt in my mind this will end up being opt-in and lazily loaded.
- When rendering with Typst, I simply overrite the `innerHTML` of the corresponding `label.element`. From what I could gather this is also what KaTeX does. However, some functionalities seem reliant on those `.katex, .katex-error` divs existing. Could you briefly explain why ? What I've noticed is that
  1. all the labels for edges have disappeared
  2. moving cells is completely broken

Do you have an idea about the possible cause of these breakages ?

Cheers

tjbcg commented:
https://github.com/varkor/quiver/blob/4fee690743fe07b37c0d57b2751ebbbae37cfb7e/src/arrow.mjs#L550

This call is returning `undefined` if I do render arrow labels with typst. The rendered version is an SVG, and in execution of `requisition_element`, [`elements.length`](https://github.com/varkor/quiver/blob/4fee690743fe07b37c0d57b2751ebbbae37cfb7e/src/arrow.mjs#L334) is actually 3. This means at some point the SVG generated by Typst was taken into account when it shouldn't have. Any idea where the corresponding selector resides ?

EDIT: Fixed, see commit below.

varkor commented:
Thanks for the pull request!

I must admit that it hadn't occurred to me that Typst has different syntax for mathematics than LaTeX, which makes support for Typst subtler. I suppose it will be necessary to have a toggle in the UI for LaTeX/Typst mode – I'll have to think about how to do that without complicating the UI too much.

- Regarding loading the binary, my feeling is that (at least for now) Typst users will be a small minority, so that it probably makes more sense to load it from a CDN (this can always be changed in the future if/when Typst becomes more popular).
- Once a KaTeX equation had been rendered, we need to examine the corresponding `.katex, .katex-error` element so that we can measure the width/height of the equation (since KaTeX does not have APIs for measuring equations, or at least did not when **quiver** was initially written). This is used to work out the bounding boxes for the label elements, so that they are positioned and cropped correctly. If there is some element whose width/height corresponds to the dimensions of a Typst equation, it should be possible to simply use that element in place of the `.katex, .katex-error` elements – I don't believe there are any special features of those elements that are being used that wouldn't be valid more generally.

Unfortunately, the timing is a little inconvenient, as I'm about to go on a research trip for a few weeks, and don't yet know whether I'll have time to look at this in detail before I get back. However, if I don't manage to look at it during my trip, it will be a priority for me once I return. I'll try to answer any questions you have in the meantime, though.

tjbcg commented:
Thanks for your time ! Regarding the second point, this was actually all addressed in the latest commit which makes me confident this is close to releasable. Don't worry about the timing though, I can just host the fork somewhere in the meantime.

What's left (for me) to do:
- [X] Colour labels correctly when rendering with typst
- [X] Add a toggle somewhere to switch between KaTeX and Typst (lazy-loading will be easy to implement). I do feel like this setting should be remembered, via localStorage most likely.
- [x] Implement the incompatibilities-reporting functionality in the typst export window, and maybe add some customization options there too

tjbcg commented:
That's it, I'm calling it releasable ! Let me know what you think.

jingkaimori commented on src/ui.mjs:
```diff
@@ -7843,7 +7967,7 @@ document.addEventListener("DOMContentLoaded", () => {
 
     // Immediately load the KaTeX library.
-   const rendering_library = new DOM.Element("script", {
+   const katex_script_dom = new DOM.Element("script", {
```
a better dynamic loader make use of `import()` rather than DOM element:
```js
    // Immediately load the KaTeX library as an ES6 module.
    KaTeX = import("/KaTeX/katex.mjs").then((module) => {
            // KaTeX is fast enough to be worth waiting for, but not
            // immediately available. In this case, we delay loading
            // the quiver until the library has loaded.
            load_quiver_from_query_string();
        return module.default;
    }).catch(() => {
        // Handle KaTeX not loading (somewhat) gracefully.
        UI.display_error("KaTeX failed to load.");
        // Remove the loading screen.
        ui.element.query_selector(".loading-screen").class_list.add("hidden");
    });
```

tjbcg replied:
While outside the scope of this PR, I've included this.

jingkaimori commented on src/ui.mjs:
```diff
@@ -7758,6 +7853,35 @@ export class Edge extends Cell {
+    render(text, template = (t) => `${CONSTANTS.TYPST_PREAMBLE}${t}`) {
+        return this.enqueue(() => new Promise(resolve => Typst.then(typst => {
+            typst.svg({mainContent: template(text)}).then(svg => resolve(svg)).catch(err => resolve(err));
+        })));
```
Some simplification here, because returning `Promise` in callback of `.then` does not results in extra wrapper:

```suggestion
        return this.enqueue(() => Typst.then(typst => {
            return typst.svg({mainContent: template(text)});
        })));
```

Besides simplification, catch clause should use second param passed by promise, usually named as "reject"

tjbcg replied:
I tried to make this work but all the simplifications in this area led to overcomplication elsewhere when calling `TypstQueue.render`. The reason is that we absolutely never want rendering errors to block the queue, meaning we can't `.catch()` on the result of `PromiseQueue.enqueue` since that catch block would yield a new promise distinct from that of `PromiseQueue.queue`. Let me know if you have a solution to this but please do some testing beforehand! Here resolving with an error (as opposed to rejecting with an error) is fine in my opinion: it's expected that most of the time an error is an acceptable outcome, since as the user is typing Typst is bound to error out along the way.

jingkaimori replied:
Returning promise from .catch clause can avoid breaking queue by exception. Error should be handled by dom render, so that visible hint(i.e. red text color like katex) could applied to label directly displayed. Detailed patch provided in another comment.

jingkaimori reviewed (commented):
Some detailed but incomplete suggestion for code

jingkaimori commented:
Some further suggestion for this code:

```patch
diff --git a/src/ui.mjs b/src/ui.mjs
index 287ff80..f8c9b20 100644
--- a/src/ui.mjs
+++ b/src/ui.mjs
@@ -5606,30 +5606,32 @@ class Panel {
         };
 
         if (ui.settings.get("quiver.renderer") === "typst") {
-            // Render the label with typst
+            // Render the label with typst. then clause must got a svg(in text), not an error
             TypstQueue.render(`$${cell.label}$`).then(result => {
                 const template = document.createElement('template');
                 template.innerHTML = result;
-                // If we got an svg and not an error
-                if (template.content.firstChild.nodeName === "svg") {
-                    const svg = template.content.firstChild;
-                    // Remove extraneous style, html text, and script, which are all generated to support
-                    // text selection in the svg (which we are not interested in)
-                    svg.querySelectorAll('script, style, foreignObject').forEach(node => node.remove())
-                    // Remove the fill attribute which prevents recolouring of the svg after the fact
-                    svg.querySelectorAll('.typst-text').forEach(node => node.removeAttribute('fill'))
-                    label.element.replaceChildren(svg);
-                    // Restore bounding box
-                    const svg_dom = label.element.children[0];
-                    const bbox = svg_dom.getBBox();
-                    const viewBox = [bbox.x, bbox.y, bbox.width, bbox.height].join(" ");
-                    svg_dom.setAttribute("viewBox", viewBox);
-                    svg_dom.setAttribute("width", bbox.width);
-                    svg_dom.setAttribute("height", bbox.height);
-                } else {
-                    // the result was an error, display the text unrendered.
-                    label.element.innerHTML = cell.label;
-                }
+                const svg = template.content.firstChild;
+                // Remove extraneous html text, which are all generated to support text selection in the svg
+                // (which we are not interested in)
+                svg.querySelectorAll('foreignObject').forEach(node => node.remove())
+                // Remove the fill attribute which prevents recolouring of the svg after the fact
+                svg.querySelectorAll('.typst-text').forEach(node => node.removeAttribute('fill'))
+                label.element.replaceChildren(svg);
+                // Restore bounding box
+                const svg_dom = label.element.children[0];
+                const bbox = svg_dom.getBBox();
+                const viewBox = [bbox.x, bbox.y, bbox.width, bbox.height].join(" ");
+                svg_dom.setAttribute("viewBox", viewBox);
+                svg_dom.setAttribute("width", bbox.width);
+                svg_dom.setAttribute("height", bbox.height);
+                update_label_transformation("typst");
+            }).catch(error => {
+                // display malformed label with style `.typst-error`, like katex.
+                // this error must be handled outside of the promise queue, because some visible hint
+                // should be provided for the user.
+                const element = new DOM.Div({ class: "typst-error" });
+                element.element.innerText = cell.label;
+                label.replace(element);
                 update_label_transformation("typst");
             });
         } else {
@@ -7862,10 +7864,7 @@ class PromiseQueue {
 
   enqueue(promiseFunction) {
     // Chain the new promise onto the existing queue
-    this.queue = this.queue.then(() => promiseFunction()).catch((error) => {
-      // Handle the error inside the queue to avoid unhandled promise rejections
-      console.error("Error executing promise:", error);
-    });
+    this.queue = this.queue.then(() => promiseFunction()).catch(() => promiseFunction());
 
     // Return the current queue state
     return this.queue;
@@ -7874,9 +7873,19 @@ class PromiseQueue {
 
 const TypstQueue = new class extends PromiseQueue {
     render(text, template = (t) => `${CONSTANTS.TYPST_PREAMBLE}${t}`) {
-        return this.enqueue(() => new Promise(resolve => Typst.then(typst => {
-            typst.svg({mainContent: template(text)}).then(svg => resolve(svg)).catch(err => resolve(err));
-        })));
+        return this.enqueue(() => Typst.then(typst => {
+            return typst.svg({
+                mainContent: template(text),
+                // Remove extraneous style, and script, which are all generated to support text selection
+                // in the svg (which we are not interested in)
+                data_selection: {
+                    body: true,
+                    defs: true,
+                    css: false,
+                    js: false,
+                }
+            });
+        }));
     }
 };
 
```

jingkaimori commented:
further suggestion:

- load typst with `import()` too
- config font loader of typst to retrieve font file from cdn

```patch
diff --git a/src/ui.mjs b/src/ui.mjs
index 31defc1..646d2b8 100644
--- a/src/ui.mjs
+++ b/src/ui.mjs
@@ -7889,8 +7889,33 @@ const TypstQueue = new class extends PromiseQueue {
     }
 };
 
-// This will be replaced on DOMContentLoaded
-let load_typst = () => console.error('Lazy loading unimplemented');
+// Load the Typst library as an ES6 module when invoked. Unlike KaTeX, this is on the heavier side of things. We don't
+// wait for it.
+const load_typst = () => {
+    Typst = import("https://cdn.jsdelivr.net/npm/@myriaddreamin/typst.ts/dist/esm/contrib/all-in-one-lite.bundle.js").then((module) => {
+        const $typst = module.$typst;
+        const preloadRemoteFonts = module.preloadRemoteFonts;
+        $typst.setCompilerInitOptions({
+            beforeBuild: [
+                preloadRemoteFonts([], {
+                    assets:['text' , 'cjk' , 'emoji'],
+                    assetUrlPrefix: "https://cdn.jsdelivr.net/gh/Myriad-Dreamin/typst@assets-fonts/"
+                })
+            ],
+            getModule: () => 'https://cdn.jsdelivr.net/npm/@myriaddreamin/typst-ts-web-compiler/pkg/typst_ts_web_compiler_bg.wasm',
+        });
+        $typst.setRendererInitOptions({
+            getModule: () => 'https://cdn.jsdelivr.net/npm/@myriaddreamin/typst-ts-renderer/pkg/typst_ts_renderer_bg.wasm',
+        });
+
+        return $typst;
+    }).catch(() => {
+        // Handle Typst.ts not loading (somewhat) gracefully.
+        UI.display_error("Typst failed to load.");
+        // Remove the loading screen.
+        ui.element.query_selector(".loading-screen").class_list.add("hidden");
+    });
+};
 
 // We want until the (minimal) DOM content has loaded, so we have access to `document.body`.
 document.addEventListener("DOMContentLoaded", () => {
@@ -7989,33 +8014,6 @@ document.addEventListener("DOMContentLoaded", () => {
         ui.element.query_selector(".loading-screen").class_list.add("hidden");
     });
 
-    const typstts_script_dom = new DOM.Element("script", {
-        type: "module",
-        src: "https://cdn.jsdelivr.net/npm/@myriaddreamin/typst.ts/dist/esm/contrib/all-in-one-lite.bundle.js",
-    }).listen("error", () => {
-        // Handle Typst.ts not loading (somewhat) gracefully.
-        UI.display_error("Typst failed to load.");
-    });
-
-
-    // Unlike KaTeX, this is on the heavier side of things. We don't wait for it.
-    Typst = new Promise((accept) => {
-        typstts_script_dom.listen("load", () => {
-            $typst.setCompilerInitOptions({
-                getModule: () => 'https://cdn.jsdelivr.net/npm/@myriaddreamin/typst-ts-web-compiler/pkg/typst_ts_web_compiler_bg.wasm',
-            });
-            $typst.setRendererInitOptions({
-                getModule: () => 'https://cdn.jsdelivr.net/npm/@myriaddreamin/typst-ts-renderer/pkg/typst_ts_renderer_bg.wasm',
-            });
-            accept($typst);
-        });
-    });
-
-    // Lazy loading function for typst.ts
-    load_typst = () => {
-        document.head.appendChild(typstts_script_dom.element);
-    }
-
     // Load immediately if Typst if the renderer set in user settings
     if (ui.settings.get("quiver.renderer") === "typst") load_typst();
 
```

jingkaimori commented:
Commit 82aaf18 introduced a bug causing typst fails to load when toggling between katex and typst mode multiple times. Here is the patch to fix it:

```patch
diff --git a/src/ui.mjs b/src/ui.mjs
index 646d2b8..e1e52d1 100644
--- a/src/ui.mjs
+++ b/src/ui.mjs
@@ -7892,6 +7892,9 @@ const TypstQueue = new class extends PromiseQueue {
 // Load the Typst library as an ES6 module when invoked. Unlike KaTeX, this is on the heavier side of things. We don't
 // wait for it.
 const load_typst = () => {
+    if (Typst !== null) {
+        return;
+    }
     Typst = import("https://cdn.jsdelivr.net/npm/@myriaddreamin/typst.ts/dist/esm/contrib/all-in-one-lite.bundle.js").then((module) => {
         const $typst = module.$typst;
         const preloadRemoteFonts = module.preloadRemoteFonts;
@@ -7912,8 +7915,6 @@ const load_typst = () => {
     }).catch(() => {
         // Handle Typst.ts not loading (somewhat) gracefully.
         UI.display_error("Typst failed to load.");
-        // Remove the loading screen.
-        ui.element.query_selector(".loading-screen").class_list.add("hidden");
     });
 };
 
@@ -8015,7 +8016,13 @@ document.addEventListener("DOMContentLoaded", () => {
     });
 
     // Load immediately if Typst if the renderer set in user settings
-    if (ui.settings.get("quiver.renderer") === "typst") load_typst();
+    if (ui.settings.get("quiver.renderer") === "typst"){
+        load_typst();
+        Typst.catch(() => {
+            // Remove the loading screen.
+            ui.element.query_selector(".loading-screen").class_list.add("hidden");
+        });
+    };
 
     // Load the style sheet needed for KaTeX.
     document.head.appendChild(new DOM.Element("link", {
```

jingkaimori commented:
`ColourPicker` class provides a LaTeX literal for color. Is it more consistent to display typst literal when typst mode is detected?

https://github.com/varkor/quiver/blob/82aaf182d1d1bd218a777cefde8b9e44eb8caf79/src/ui.mjs#L6901-L6904

tjbcg commented:
> `ColourPicker` class provides a LaTeX literal for color. Is it more consistent to display typst literal when typst mode is detected?
> 
> https://github.com/varkor/quiver/blob/82aaf182d1d1bd218a777cefde8b9e44eb8caf79/src/ui.mjs#L6901-L6904

This list is only ever filled up when importing LaTeX macros and when those macros contain shorthands for color definitions (if I read the code correctly). I guess it would make sense to hide this list when in Typst mode, as there is currently no way to import Typst macros (although this wouldn't be too hard to implement since we use the actual typst compiler with a predefined preamble).

A note for implementation if going in that direction: in a macro list, we can retrieve variable names via standard matching (e.g. `/^#let\s+(\w+)\s*=/`) and then filter out which of these variables are colors with typst `type(var) == "color"`.

I say in the current state of things, this list can remain hidden in Typst mode. I'm also wondering if it's not worth hiding the Typst export button when in LaTeX mode. This would allow in the future to add buttons for fletcher import and typst macro import without having too much clutter in the bottom bar. Exporting a fletcher diagram from a LaTeX one will also almost always yield incorrect maths syntax (although that can be dealt with using the typst package `mitex`).

jingkaimori commented:
IMO importing external typst content could be implemented in another pr.

> I'm also wondering if it's not worth hiding the Typst export button when in LaTeX mode.

I agree with this. And mode toggling button could be moved to bottom bar, so that things related to label rendering appears together.

<details><summary>Solved content</summary>
<p>

----

Default mode applied to input is differ between KaTeX and typst. In KaTeX, input is considered as if in math mode, but in typst, input is treated as normal text.

In this pr, a wrapper is applied automatically when typst is selected as renderer. Therefore changing label to text mode needs unnecessary wrapper.

https://github.com/varkor/quiver/blob/82aaf182d1d1bd218a777cefde8b9e44eb8caf79/src/quiver.mjs#L325
https://github.com/varkor/quiver/blob/82aaf182d1d1bd218a777cefde8b9e44eb8caf79/src/ui.mjs#L5610

But according to https://github.com/varkor/quiver/issues/45#issuecomment-734014878 such wrapper should be applied by user. 
A better solution is rendering text without math mode wrapping `"$${text}$"`, and create new node with initial label `"$$"`

</p>
</details>

Myriad-Dreamin commented on src/ui.mjs:
```diff
@@ -7758,6 +7855,67 @@ export class Edge extends Cell {
+// Load the Typst library as an ES6 module when invoked. Unlike KaTeX, this is on the heavier side of things. We don't
+// wait for it.
+const load_typst = () => {
+    Typst = import("https://cdn.jsdelivr.net/npm/@myriaddreamin/typst.ts/dist/esm/contrib/all-in-one-lite.bundle.js").then((module) => {
```
Hello, I'm the owner of typst.ts.

I notice that you didn't lock the version to use here, which is considered not good for stability. All the imported libraries should be locked to one version.

varkor commented:
I'm now back from my trip and plan to review this soon. Feel free to ping me in a couple of weeks if I haven't gotten round to it by then.

RoyPrinsGH commented:
@varkor Bump

varkor commented:
Hey, I've taken a look at the pull request (apologies for this taking so long – there will be less delay in the future). You've made good progress. A few general comments, without having looked at the code yet:

- The first time a new object is created (e.g. by double-clicking) in Typst mode, it takes a little while for the label to render. After that, labels seem to render quickly. Is there a way to eliminate this delay, e.g. by preloading something?
- It would be good to only have a single button to export to LaTeX/Typst that depends on the mode. (Perhaps the simplest solution is to hide the LaTeX or Typst buttons depending on the current mode, rather than trying to make a single button context-dependent.)
- The error styling CSS for Typst syntax errors should be the same as for KaTeX errors.
- At the moment, the user is expected to include `$ ... $` to render labels as mathematics. I think it would make sense to make this implicit, so that it matches the behaviour of LaTeX (unless there's a good reason not to do this).
- It would be ideal for **quiver** URLs to record the kind of diagram, so that when opening a URL, we know which mode to load into. However, this will require changing the URL format, so I'll need to think about the best way to do this. The simplest strategy would be to add a new query parameter, rather than modifying the diagram metadata.

Regarding some of the earlier comments in the discussion:
- I think I need to reconsider how macros are handled, so you can just ignore this for now.
- I may play around a little bit with the UI for switching between modes, so it's fine to leave in the toolbar for now.

tjbcg commented:
I haven't got much time in the next few weeks (enough to answer here, not enough to actually write code)  but I'll be able to take a look afterwards.

Regarding your first point, it is possible to preload the Typst compiler but I decided against it because of its size (IIRC around 20MB). Up to you!

Agreed with all your other points.

Edit to add: regarding macros, they can be handled transparently in Typst since the full compiler is brought in and each node is actually rendered as a full typst document.

varkor commented:
Sorry, I somehow completely missed the notification for the new message.

> Regarding your first point, it is possible to preload the Typst compiler but I decided against it because of its size (IIRC around 20MB). Up to you!

I don't think we should preload the compiler by default when loading the page. However, I think it would be reasonable to preload if the Typst query parameter is present; and as soon as the user clicks on the Typst mode button.

tjbcg commented:
Hi there ! Finally found time to work on this. I've pushed something that addresses most of your points from https://github.com/varkor/quiver/pull/262#issuecomment-2907758110

I've reworked the renderer-related UI: the button to change renderer moved away from the top bar and into the bottom panel as a dropdown. Buttons in this panel are shown or hidden depending on the current renderer. I think it makes much more sense this way (input/output related functions at the bottom, content/diagram related functions at the top).

About the initial rendering delay in typst: unfortunately this is not fixable for the forseeable future, because this startup delay is actually the startup time of the typst compiler as bundled in typst.ts (that is, it's the startup time of the wasm binary, more or less).

All of your other points have been addressed. Let me know what you think!

varkor commented:
Thank you! I will take a look in the upcoming week!

varkor commented:
I think this is very close to being ready. My remaining concern is about the initial rendering delay. My recollection is that the last time I tested the pull request, the initial rendering delay was only about a second, but now it seems to be be taking several seconds, so I think we need to do whatever we can to reduce the delay the user sees.

(I would guess that a significant portion of the startup time of the Typst compiler corresponds to things that are irrelevant for mathematics rendering, in which case it would be nice if it was possible to load a more minimal compiler with a reduced startup time, but presumably this isn't possible.)

> About the initial rendering delay in typst: unfortunately this is not fixable for the forseeable future, because this startup delay is actually the startup time of the typst compiler as bundled in typst.ts (that is, it's the startup time of the wasm binary, more or less).

At the very least, presumably one way to reduce the startup delay would be to render something (without adding it to the DOM) as soon as the Typst renderer is chosen using the drop-down menu, or as soon as the page loads if it loads in Typst mode?

I also wonder whether it might be more user-friendly to display plain text labels until the Typst compiler has loaded, so the user sees that what they're typing isn't being ignored?

Furthermore, when a diagram is loaded in Typst mode via a URL, I think the loading screen should not be dismissed until the diagram has been fully rendered. Otherwise, the page loads to a bunch of arrows without any labels.

tjbcg commented:
I don't think the rendering delay changed (the library version is pinned and I haven't changed it). I've implemented what you proposed:
- Upon loading a diagram via query string with typst, the loading screen is shown until labels are rendered
- When loading the library, an empty element is rendered in the background to actually load the compiler in memory
- When typst is being loaded, labels are rendered as plaintext with a light css animation to show that it's loading (the animation is a simple fade loop between black and gray).

Let me know how that looks!

varkor commented:
@tjbcg: thanks, this looks great! I think I'm happy with everything now, so if you could rebase to address the merge conflicts and remove the Vim settings comment, I'll go ahead and merge. I'll probably end up making a few minor tweaks, but I hope to push a new version of **quiver** with these changes in the next couple of weeks. Thanks for all of your hard work (and patience with my slow responses)!

tjbcg commented:
Done! Thank you.

varkor commented:
(I'm going to make some changes to a local copy of the pull request in the next few days, and then merge once I have made a few tweaks.)

varkor committed:
Implement Typst support

Co-Authored-By: Théophile Cailliau <8740385+tjbcg@users.noreply.github.com>

varkor commented:
I did some minor tweaking and refactoring and fixed a couple of tiny issues with the fletcher output. I'm satisfied with how everything is working now (I'm sure as people start using the feature there will be other things to fix, but I imagine nothing major). I'm going to merge this now. I'm going to take the opportunity to address a few issues that have recently arisen with the latest version of Safari, and then I'll push a new version of **quiver** (hopefully by the end of the week) including Typst support.

Thank you again very much @tjbcg for implementing this feature!

lucianchauvin commented:
Great work! I am really excited to get to use Quiver again!

varkor commented:
This is now live at https://q.uiver.app/.
